app/src/main/java/org/mozilla/fenix/settings/creditcards/CreditCardValidator.kt

```diff
@@ -0,0 +1,57 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.settings.creditcards
+
+import androidx.annotation.StringRes
+import mozilla.components.concept.storage.CreditCard
+import mozilla.components.concept.storage.CreditCardNumber
+import org.mozilla.fenix.R
+
+class CreditCardValidator(
+    private val creditCardsList: List<CreditCard>
+) {
+
+    /**
+     * Validates the credit card information entered by the user.
+     * @return a [ValidationResult] that describes the validation result,
+     * either a [ValidationResult.Valid] or a [ValidationResult.Invalid]
+     */
+    internal fun validateCreditCard(creditCardNumber: String, editing: Boolean): ValidationResult {
+        if (creditCardNumber.validateCreditCardNumber()) {
+            if (editing && checkForDuplicates(creditCardNumber)) {
+                return ValidationResult.Invalid(R.string.credit_cards_number_duplication_error_message)
+            }
+        } else {
+            return ValidationResult.Invalid(R.string.credit_cards_number_validation_error_message)
+        }
+
+        return ValidationResult.Valid()
+    }
+
+    /**
+     * Validates the credit card number against existing cards list
+     * @return true if the credit card is a duplicate, false otherwise.
+     */
+    private fun checkForDuplicates(creditCardNumber: String): Boolean {
+
+        val creditCardsNumberList: List<CreditCardNumber.Encrypted> = creditCardsList.map {
+            it.encryptedCardNumber
+        }
+
+        if (CreditCardNumber.Encrypted(creditCardNumber) in creditCardsNumberList) {
+            return true
+        }
+
+        return false
+    }
+
+    /**
+     * A class that describes a credit card validation result.
+     */
+    sealed class ValidationResult {
+        data class Invalid(@StringRes val value: Int) : ValidationResult()
+        data class Valid(val value: Boolean = true) : ValidationResult()
+    }
```

Comment:
Seeing this I am wondering if we should depend on https://github.com/mozilla-mobile/android-components/pull/9875 which will tell you whether or not a credit can be added or updated.

Reply:
Maybe I will halt this until that is finished and merged. At the moment I do not see any input validation in `CreditCardValidationDelegate`, just info about the possibility of updating or saving a credit card. Is there any flow where new credit cards are saved in other places than `CreditCardEditorFragment`? In a prompt or something?

Reply:
@gabrielluong I am thinking of keeping this PR for just handling the extraction of the validator so we can easily add other validation for date and name if UX/product decides to do this.
For the duplication part, to avoid the need to pass the credit card list to this validator, do you think the check can be handled by the validator you are working on in AC, or should the save/ update call to storage return a response that we can handle in Fenix? Can we enforce the no duplicates rule in the database or do we depend on a-s implementations?

Reply:
I think we should ignore the validator I am working on in AC. We should handle duplicates here as well this is actually a P1 priority from product. Looking into the Logins implementation of this, I think they do fetch the list of logins from the DB and have some particular AS functions that we don't have. I don't think we should rely on the list from the state. 

We may need to ask for some AS functions as needed.

I think my AC validator will only used for the prompts, but since I no longer see how my validator fits here, I actually halted work on it since we don't have the save/update prompts which would've utilized it.